src/test/unit_test/multi_runner.h

```diff
@@ -0,0 +1,354 @@
+    std::size_t failed = 0;
+    typename clock_type::time_point start = clock_type::now();
+
+    explicit
+    suite_results(std::string name_ = "")
+        : name(std::move(name_))
+    {
+    }
+
+    void
+    add(case_results const& r);
+};
+
+struct results
+{
+    using static_string = beast::static_string<256>;
+    // results may be stored in shared memory. Use `static_string` to ensure
+    // pointers from different memory spaces do not co-mingle
+    using run_time = std::pair<static_string, typename clock_type::duration>;
+
+    enum { max_top = 10 };
+
+    std::size_t suites = 0;
+    std::size_t cases = 0;
+    std::size_t total = 0;
+    std::size_t failed = 0;
+    boost::container::static_vector<run_time, max_top> top;
+    typename clock_type::time_point start = clock_type::now();
+
+    void
+    add(suite_results const& r);
+
+    void
+    add(results const& r);
+
+    template <class S>
+    void
+    print(S& s);
+};
+
+template <bool IsParent>
+class multi_runner_base
+{
+    // `inner` will be created in shared memory. This is one way
+    // multi_runner_parent and multi_runner_child object communicate. The other
+    // way they communicate is through message queues.
+    struct inner
+    {
+        std::atomic<std::size_t> job_index_{0};
+        std::atomic<std::size_t> test_index_{0};
+        std::atomic<bool> any_failed_{false};
+        // A parent process will periodically increment `keep_alive_`. The child
+        // processes will check if `keep_alive_` is being incremented. If it is
+        // not incremented for a sufficiently long time, the child will assume the
+        // parent process has died.
+        std::atomic<std::size_t> keep_alive_{0};
+
+        mutable boost::interprocess::interprocess_mutex m_;
+        detail::results results_;
+
+        std::size_t
+        checkout_job_index();
+
+        std::size_t
+        checkout_test_index();
+
+        bool
+        any_failed() const;
+
+        void
+        any_failed(bool v);
+
+        void
+        inc_keep_alive_count();
+
+        std::size_t
+        get_keep_alive_count();
+
+        void
+        add(results const& r);
+
+        template <class S>
+        void
+        print_results(S& s);
+    };
+
+    static constexpr const char* shared_mem_name_ = "RippledUnitTestSharedMem";
+    // name of the message queue a multi_runner_child will use to communicate with
+    // multi_runner_parent
+    static constexpr const char* message_queue_name_ = "RippledUnitTestMessageQueue";
+
+    // `inner_` will be created in shared memory
+    inner* inner_;
+    // shared memory to use for the `inner` member
+    boost::interprocess::shared_memory_object shared_mem_;
+    boost::interprocess::mapped_region region_;
+
+protected:
+    std::unique_ptr<boost::interprocess::message_queue> message_queue_;
+
+    void message_queue_send(std::string const& s);
+
+public:
+    multi_runner_base();
+    ~multi_runner_base();
+
+    std::size_t
```

Comment:
No, all the members are not used for both. For example, parent process never `checkout_job_index`, and child processes never `inc_keep_alive_count`. However, I don't want to move those functions to the leaf classes. The rational is `multi_runner_base` puts everything that deals with shared memory in one place, I don't want to scatter that between the parent and child classes.

Let me try to address the why I went with inheritance. So the two choices I see are making it a member and using private inheritance. The `multi_runner_base` was originally `shared_state` and packaged up everything that was stored in shared memory. The parent and child had data members of this type (no inheritance). The problem came when I had to serialize access to the message queue on OSX. The message queue does _not_ live in shared memory, but needs to use a mutex that _does_ live in shared memory. I want the message queue to be a member of the class (the code gets ugly otherwise - even when using pass through functions). So instead of making this a member, I used private inheritance. There are certainly other ways of doing the same thing, but I still think private inheritance is a good choice for this.